sp/src/vscript/sqdbg/sqdbg/json.h

```diff
@@ -0,0 +1,1496 @@
+	{
+		case JSON_STRING:
+		{
+			FreeString( &val->_string );
+			break;
+		}
+#ifdef SQUNICODE
+		case JSON_WSTRING:
+		{
+			FreeString( &val->_wstring );
+			break;
+		}
+#endif
+		case JSON_TABLE:
+		{
+			DeleteTable( val->_table );
+			break;
+		}
+		case JSON_ARRAY:
+		{
+			DeleteArray( val->_array );
+			break;
+		}
+		default: UNREACHABLE();
+	}
+}
+
+inline int GetJSONStringSize( const json_value_t &obj )
+{
+	int len;
+
+	switch ( obj.GetType() )
+	{
+		case JSON_STRING:
+		{
+			len = 2 + obj._string.len; // "val"
+
+			if ( obj.IsType( _JSON_QUOTED ) )
+				len += 4;
+
+			const char *c = obj._string.ptr;
+			for ( int i = obj._string.len; i--; c++ )
+			{
+				switch ( *c )
+				{
+					case '\"': case '\\': case '\b':
+					case '\f': case '\n': case '\r': case '\t':
+						len++;
+						if ( obj.IsType( _JSON_QUOTED ) )
+						{
+							len++;
+							if ( *c == '\"' || *c == '\\' )
+								len++;
+						}
+						break;
+					default:
+						if ( !IN_RANGE_CHAR( *(unsigned char*)c, 0x20, 0x7E ) )
+						{
+							int ret = IsValidUTF8( (unsigned char*)c, i + 1 );
+							if ( ret != 0 )
+							{
+								i -= ret - 1;
+								c += ret - 1;
+							}
+							else
+							{
+								if ( !obj.IsType( _JSON_QUOTED ) )
+								{
+									len += sizeof(uint16_t) * 2 + 1;
+								}
+								else
+								{
+									len += sizeof(SQChar) * 2 + 2;
+								}
+							}
+						}
+				}
+			}
+
+			break;
+		}
+#ifdef SQUNICODE
+		case JSON_WSTRING:
+		{
+			if ( !obj.IsType( _JSON_QUOTED ) )
+			{
+				len = 2 + UTF8Length< kUTFEscapeJSON >( obj._wstring.ptr, obj._wstring.len );
+			}
+			else
+			{
+				len = 2 + UTF8Length< kUTFEscapeQuoted >( obj._wstring.ptr, obj._wstring.len );
+			}
+
+			break;
+		}
+#endif
+		case JSON_INTEGER:
+		{
+			if ( obj._integer > 0 )
+			{
+				len = countdigits( obj._integer );
+			}
+			else if ( obj._integer )
+			{
+				len = countdigits( -obj._integer ) + 1;
+			}
+			else
+			{
+				len = 1; // 0
+			}
+
+			if ( obj.IsType( _JSON_QUOTED ) )
+				len += 2;
+
+			break;
+		}
+		case JSON_FLOAT:
+		{
+			char tmp[ FMT_FLT_LEN + 1 ];
+			len = snprintf( tmp, sizeof(tmp), "%f", obj._float );
+			break;
+		}
+		case JSON_BOOL:
+		{
+			len = obj._integer ? STRLEN("true") : STRLEN("false");
+			break;
+		}
+		case JSON_NULL:
+		{
+			len = STRLEN("null");
+			break;
+		}
+		case JSON_TABLE:
+		{
+			len = 2; // {}
+
+			if ( obj._table->size() )
+			{
+				for ( int i = 0; i < obj._table->size(); i++ )
+				{
+					json_field_t *kv = obj._table->Get(i);
+					len += 2 + kv->key.len + 1 + GetJSONStringSize( kv->val ) + 1; // "key":val,
+				}
+
+				len--; // trailing comma
+			}
+
+			break;
+		}
+		case JSON_ARRAY:
+		{
+			len = 2; // []
+
+			if ( obj._array->size() )
+			{
+				for ( int i = 0; i < obj._array->size(); i++ )
+					len += GetJSONStringSize( *obj._array->Get(i) ) + 1;
+
+				len--; // trailing comma
+			}
+
+			break;
+		}
+		default: UNREACHABLE(); GCC_BLOCK( len = 0; );
```

Comment:
It seems the [`UNREACHABLE()` macro](https://github.com/mapbase-source/source-sdk-2013/blob/471a840ed98c7206237cb579671a6d6fda9fd4f9/sp/src/public/tier0/platform.h#L607) is broken for GCC. It relies on [`HINT()`](https://github.com/mapbase-source/source-sdk-2013/blob/471a840ed98c7206237cb579671a6d6fda9fd4f9/sp/src/public/tier0/platform.h#L602), which is unimplemented for gcc. It should be `#define HINT(THE_HINT) while { if ( ! ( THE_HINT ) ) __builtin_unreachable(); } while (0)`. That should make the `GCC_BLOCK`s unecessary.

Reply:
I should've guessed that when my local gcc compile worked but not the Source compile. The only reason `tier0/dbg.h` is included is because CrtDbg functions are stubs in Source instead of calling the custom assertion functions.

I think I'll just `#if 0` assert definitions instead of the entire `debug.h` file and undef Valve's `UNREACHABLE`.